compiler/rustc_builtin_macros/src/concat_bytes.rs

```diff
@@ -0,0 +1,210 @@
+use rustc_ast as ast;
+use rustc_ast::tokenstream::TokenStream;
+use rustc_data_structures::sync::Lrc;
+use rustc_errors::Applicability;
+use rustc_expand::base::{self, DummyResult};
+
+pub fn expand_concat_bytes(
+    cx: &mut base::ExtCtxt<'_>,
+    sp: rustc_span::Span,
+    tts: TokenStream,
+) -> Box<dyn base::MacResult + 'static> {
+    let es = match base::get_exprs_from_tts(cx, sp, tts) {
+        Some(e) => e,
+        None => return DummyResult::any(sp),
+    };
+    let mut accumulator = Vec::new();
+    let mut missing_literal = vec![];
+    let mut has_errors = false;
+    for e in es {
+        match e.kind {
+            ast::ExprKind::Array(ref exprs) => {
+                for expr in exprs {
+                    match expr.kind {
+                        ast::ExprKind::Array(_) => {
+                            cx.span_err(expr.span, "cannot concatenate doubly nested array");
+                            has_errors = true;
+                        }
+                        ast::ExprKind::Lit(ref lit) => match lit.kind {
+                            ast::LitKind::Str(_, _) => {
+                                cx.span_err(
+                                    expr.span,
+                                    "cannot concatenate an array of string literals",
+                                );
+                                has_errors = true;
+                            }
+                            ast::LitKind::Float(_, _) => {
+                                cx.span_err(
+                                    expr.span,
+                                    "cannot concatenate an array of float literals",
+                                );
+                                has_errors = true;
+                            }
+                            ast::LitKind::Char(_) => {
+                                if let Ok(snippet) = cx.sess.source_map().span_to_snippet(expr.span)
+                                {
+                                    cx.struct_span_err(
+                                        expr.span,
+                                        "cannot concatenate an array of character literals",
+                                    )
+                                    .span_suggestion(
+                                        expr.span,
+                                        "try using a byte character",
+                                        format!("b{}", snippet),
+                                        Applicability::MachineApplicable,
+                                    )
+                                    .emit();
+                                } else {
+                                    cx.span_err(
+                                        expr.span,
+                                        "cannot concatenate an array of character literals",
+                                    )
```

Comment:
Can we avoid duplicating the error creation across this if? We should be able to give the suggestion only if we have the snippet, I think.